cluster-alerts: Add alert rule DeprecatedMachineType

dasionov opened the pull request:
**What this PR does / why we need it**:

This PR adds the `DeprecatedMachineType` alert to detect when a VM is using a deprecated machine type.

Rule Expression:
```
kubevirt_vm_info
* on(machine_type) group_left()
  max(kubevirt_node_deprecated_machine_types) by (machine_type)
```

How It Works:
- Uses only `kubevirt_vm_info`, covering both running and stopped VMs.
- Joins each VM's machine type against the cluster's deprecated types.
- Fires when any VM uses a deprecated type.

Alert Details:
- Fires after 5m of continuous match.
- Summary: VM <namespace>/<name> is using a deprecated machine type.
- Description: Includes VM name, namespace, and machine type.
- Severity: warning, operator_health_impact: warning.

Why this is needed:
- Provides early visibility into workloads running on machine types that will be removed in future releases.
- Allows administrators to proactively migrate workloads to supported types before upgrading clusters, reducing the risk of downtime or incompatibility.

**Reviewer Checklist**

- [ ] PR Message
- [ ] Commit Messages
- [ ] How to test
- [ ] Unit Tests
- [ ] Functional Tests
- [ ] User Documentation
- [ ] Developer Documentation
- [ ] Upgrade Scenario
- [ ] Uninstallation Scenario
- [ ] Backward Compatibility
- [ ] Troubleshooting Friendly

**Jira Ticket**:
```jira-ticket
https://issues.redhat.com/browse/CNV-56660
```

**Release note**:
```release-note
add new alert `DeprecatedMachineType` for detecting vms with deprecated machine types.
```

dasionov commented:
/hold, wait for https://github.com/kubevirt/kubevirt/pull/14255 to introduce the metric.

sradco commented on pkg/monitoring/observability/rules/alerts/cluster_alerts.go:
```diff
@@ -59,5 +59,18 @@ func clusterAlerts() []promv1.Rule {
+			},
+			Labels: map[string]string{
+				"severity":               "warning",
+				"operator_health_impact": "none",
```
Isn't this a blocker for an upgrade? Shouldnt we degrade the operator and set `operator_health_impact `to `warning`?

dasionov commented:
[Change](https://github.com/kubevirt/hyperconverged-cluster-operator/pull/3358/commits/df09099c10df14705cb5138e3e3e8727633da036): Rebase

dasionov commented:
/test pull-hyperconverged-cluster-operator-e2e-k8s-1.33

dasionov commented:
/retest-required

dasionov commented:
/test all

nunnatsa commented on tests/func-tests/monitoring_test.go:
```diff
@@ -172,6 +174,74 @@ var _ = Describe("[crit:high][vendor:cnv-qe@redhat.com][level:system]Monitoring"
+
+		It("when a VM is using a deprecated machine type", func(ctx context.Context) {
+			By("Creating a VM with a deprecated machine type")
+			runStrategy := kubevirtcorev1.RunStrategyAlways
+			vm := &kubevirtcorev1.VirtualMachine{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "deprecated-vm",
+					Namespace: tests.TestNamespace,
+				},
+				Spec: kubevirtcorev1.VirtualMachineSpec{
+					RunStrategy: &runStrategy,
```
Consider using `ptr.To` instead of defining the `runStrategy` variable; e.g.
```suggestion
			vm := &kubevirtcorev1.VirtualMachine{
				ObjectMeta: metav1.ObjectMeta{
					Name:      "deprecated-vm",
					Namespace: tests.TestNamespace,
				},
				Spec: kubevirtcorev1.VirtualMachineSpec{
					RunStrategy: ptr.To(kubevirtcorev1.RunStrategyAlways),
```

nunnatsa reviewed (commented):
Thanks @dasionov.

In the functional test, please make sure the VM and the AMI are deleted when the test  exits.

Aloso, added one nit inline comment.

dasionov commented:
> Thanks @dasionov.
> 
> In the functional test, please make sure the VM and the AMI are deleted when the test exits.
> 
> Also, added one nit inline comment.

Added `DeferCleanup` call for both VM and VMI test, waiting for https://github.com/kubevirt/monitoring/pull/305 to be merged before moving this PR to ready

dasionov commented on pkg/monitoring/observability/rules/alerts/cluster_alerts.go:
```diff
@@ -133,5 +133,38 @@ func clusterAlerts() []promv1.Rule {
+				  label_replace(kubevirt_vmi_info, "machine_type", "$1", "guest_os_machine", "(.*)")
+			   )
+			   * on(machine_type) group_left
+			   max(kubevirt_node_deprecated_machine_types) by (machine_type)
```
Fixed the expression as you suggested, not sure i understand what you meant about the status thing, could you elaborate?

sradco commented on pkg/monitoring/observability/rules/alerts/cluster_alerts.go:
```diff
@@ -133,5 +133,22 @@ func clusterAlerts() []promv1.Rule {
+			`),
+			For: ptr.To(promv1.Duration("5m")),
+			Annotations: map[string]string{
+				"summary":     "VM {{ $labels.namespace }}/{{ $labels.name }} is using a deprecated machine type",
```
This is not clear.
I suggest updating this to:
Virtual Machine {{ $labels.name }}  in namespace {{ $labels.namespace }} is using a deprecated machine type.

Same for the description.

dasionov replied:
fixed

sradco commented on hack/prom-rule-ci/observability-prom-rules-tests.yaml:
```diff
@@ -417,3 +417,27 @@ tests:
+              name: "deprecated-vm"
+              machine_type: "pc-q35-rhel7.6.0"
+              kubernetes_operator_part_of: "kubevirt"
+              kubernetes_operator_component: "cnv-observability"
```
missing runbook_url="https://kubevirt.io/monitoring/runbooks/DeprecatedMachineType"

dasionov replied:
fixed

sradco commented on hack/prom-rule-ci/observability-prom-rules-tests.yaml:
```diff
@@ -417,3 +417,27 @@ tests:
+              description: "Virtual Machine 'deprecated-vm' in namespace 'default' is using machine type 'pc-q35-rhel7.6.0', which is deprecated."
+            exp_labels:
+              severity: "warning"
+              operator_health_impact: "warning"
```
Does this really impact the operator health? do you expect the operator do be "degraded" if this alert fires for specific VM?

dasionov replied:
no, it should be none

nunnatsa commented:
/hold

Should we wait with it for the next version, @dasionov

nunnatsa commented:
/unhold

sradco commented on hack/prom-rule-ci/observability-prom-rules-tests.yaml:
```diff
@@ -417,3 +417,28 @@ tests:
+  # Test for DeprecatedMachineType alert
+  - interval: 1m
+    input_series:
+      - series: 'kubevirt_vm_info{name="deprecated-vm", namespace="default", machine_type="pc-q35-rhel7.6.0"}'
```
You need to add the status_group to the metric here

dasionov commented:
/retest-required

dasionov commented:
runbook PR: https://github.com/kubevirt/monitoring/pull/305

nunnatsa commented:
/override pull-hyperconverged-cluster-operator-unit-test-s390x

The s390x test server is broken.
